Improve layout of documentation page and add search feature

Inverle opened the pull request:
Closes https://github.com/FreshRSS/FreshRSS/issues/7915, https://github.com/FreshRSS/FreshRSS/issues/5325

Demo: https://inverle.github.io/FreshRSS/

Inverle committed:
Improve layout of documentation page and add search feature

Also: anchor headings and fix building site locally

Inverle commented on docs/Gemfile:
Needed for proper working of local build

Inverle commented on docs/_config.yml:
```diff
@@ -1,9 +1,19 @@
-exclude: [CHANGELOG*.md]
+exclude: [CHANGELOG*.md, README.md]
+
+plugins:
```
Needed for proper working of local build

Frenzie commented:
You're probably aware, but the mobile styles seem to kick in too late (should be at something like 1200 already, also compare [Bootstrap](https://getbootstrap.com/docs/5.3/getting-started/introduction/)).

<kbd><img width="1202" height="2144" alt="image" src="https://github.com/user-attachments/assets/a0d9ac80-1043-48e8-b98d-620adb1d4d1c" /></kbd>

Inverle committed:
Further improvements

* Set color of hyperlinks
* Consistent styling of close aside button across devices
* Mobile layout 600px -> 1200px
* Add suffix to docs `<title>`
    * Note: titles of pages probably need to be improved, since currently they are just derived from the names of the first heading on every page
* Add favicon

Inverle committed:
Improve font

Inverle commented on docs/assets/css/style.css:
```diff
@@ -0,0 +1,137 @@
+@import url('https://fonts.googleapis.com/css2?family=Inconsolata:wght@200..900&display=swap');
```
Still need to improve, because a different font is used before this one is loaded and that makes page loads look weird

Inverle committed:
Try to fix favicon not loading correctly on GH pages

Inverle committed:
Use local font

Inverle committed:
Attempt to fix GH pages

Inverle committed:
Final improvements

* Copy to clipboard button
* Support for nojs search
* Dark mode
* Load search.json (200KB json) only on search input focus
* Keep scroll state of sidebar across navigations

Inverle committed:
Clickable images and CSP

CSP so we avoid hotlinking resources and clickable images are useful for zooming on mobile for example

Inverle commented:
Please review again

Inverle committed:
Fix typos

Inverle committed:
Disable Dark Reader extension if dark mode CSS is loaded

Alkarex commented:
Nice work, looks good to me 👍🏻 
My main feedback would be on internationalisation: the root seems to auto-redirect to `/en/`. It would be nice to be able to see that there is both English and French.
And a similar comment about search, which seems to search in both languages, which is not necessary a problem (it might even be good). If possible, just adding a horizontal separator in the search results between the languages could be enough:

<img width="468" height="376" alt="image" src="https://github.com/user-attachments/assets/8faa81c0-4579-4faf-9d6b-5b085333da35" />

Frenzie commented on docs/index.md:
```diff
@@ -4,3 +4,6 @@ If you want to contribute, you can [find us on GitHub](https://github.com/FreshR
 - [Documentation française](./fr/index.md)
+
+<!-- todo: french docs -->
+<meta http-equiv="Refresh" content="0; url=./en/">
```
You're probably aware, but you can match against navigator.languages. Something in this general direction.

```js
const supportedLanguages = ['en', 'fr'];
// supported for a decade, probably not worth checking anything besides a try/catch just in case
for (const lang of navigator.languages) {
 const normalizedLang = lang.toLowerCase();
 const matched = supportedLanguages.find(supported =>
  supported.toLowerCase() === normalizedLang ||
  normalizedLang.startsWith(supported + '-')
 );
 location.href = ...
 break; // script can sometimes keep running for a couple hundred ms more; it'd be weird to redirect to the second choice
}
```

I wonder if the refresh `<meta>` works in a `<noscript>`?

Inverle replied:
This will be replaced by a language dropdown instead, it is probably not a good idea to auto redirect to French, since it's not as up to date as the English documentation

Inverle commented:
I have a WIP language dropdown

Inverle committed:
Support internationalisation (via language dropdown)

Inverle committed:
Add Gemfile.lock

Inverle committed:
Make CI build work with the custom plugin

Inverle commented:
@Alkarex See https://inverle.github.io/FreshRSS/

Inverle commented on docs/_config.yml:
```diff
@@ -1,9 +1,53 @@
+    close: Close
+    choose_language: Choose language
+    copy_to_clipboard: Copy to clipboard
+  fr:
+    back_to_freshrss: Retour à freshrss.org
+    search_docs: Rechercher dans la documentation…
+    search: Rechercher
+    toggle_aside: Afficher/masquer le panneau
+    close: Fermer
+    choose_language: Choisir la langue
+    copy_to_clipboard: Copier dans le presse-papiers
```
This is translated with ChatGPT so correct if needed

Frenzie replied:
It's correct according to https://fr.techdico.com/traduction/anglais-francais/copy+to+clipboard.html

Alkarex replied:
Looks good

Inverle commented on docs/_includes/docs_nav.html:
```diff
@@ -0,0 +1,77 @@
+	<ul>
+		<li><a href="{{ '/fr/' | relative_url }}">Accueil</a></li>
+		<li>
+			<a href="{{ '/fr/users/02_First_steps.html' | relative_url }}">Documentation utilisateur</a>
```
Same here and below

Inverle commented:
What should be done here? (typos CI fail)
```
error: `contributeurs` should be `contributes`
  --> ./docs/_includes/docs_nav.html:73:78
   |
73 |    <a href="{{ '/fr/contributing.html' | relative_url}}">Directives pour les contributeurs</a>
   |                                                                              ^^^^^^^^^^^^^
   |
```

Inverle committed:
Make menus closable with Esc

Frenzie commented:
> (typos CI fail)

Just yesterday I named something `anc_whatever` (for ancestor_whatever) and a similar system complained it should be `and`. I caved and just called it `ancestor_whatever` since there was an easy solution. :-)

As for the question I'd say just exclude all of French? Some kind of English spellcheck seems extremely unlikely to be useful there. ;-)

Inverle committed:
Fix typos CI

Inverle committed:
Merge branch 'edge' into docs-search

math-GH commented:
I like it to have the navigation on the left hand side.

But I am not a fan of this general layout. It looks very nerdy and difficult to read. Is there any chance of a more conservative style?

Inverle commented:
> But I am not a fan of this general layout. It looks very nerdy and difficult to read. Is there any chance of a more conservative style?

Difficult to read because of the font? I thought about changing it, but not sure to which one.
As for the layout, I'm also open to suggestions.

math-GH commented:
That are the fonts of Origine theme: "OpenSans", "Cantarell", "Helvetica", "Arial"

Alkarex commented on .github/workflows/jekyll-gh-pages.yml:
```diff
@@ -31,10 +31,11 @@ jobs:
-          destination: ./_site
+        run: |
+          cd docs
+          sudo gem install bundler
```
I have not looked in details, but `sudo` looks strange to me there. Might be installable locally instead of globally?

Inverle replied:
Yes, there is an option for that I think, but doesn't make a difference here.

Inverle replied:
Though I probably should mention that this step increases the build time in CI from 19s to 59s.

https://github.com/FreshRSS/FreshRSS/actions/runs/19749114713/job/56588969757
vs
https://github.com/Inverle/FreshRSS/actions/runs/19775210787/job/56666596731

Alkarex replied:
Any chance of using caching some more?
`gem install bundler` should not run every time

Inverle replied:
I just found https://github.com/actions/jekyll-build-pages/issues/46#issuecomment-1215505246, will look into improving this. (you're correct about that)

Inverle replied:
Okay it now takes 35s to build when gems are already cached:

https://github.com/Inverle/FreshRSS/actions/runs/19804796794/job/56737317095
https://github.com/Inverle/FreshRSS/actions/caches

Inverle replied:
It's still a long build time, but I think the build got slower after I added either the `jekyll-i18n_tags` plugin or

https://github.com/FreshRSS/FreshRSS/blob/6e87494b4da378287d169e550b9211eeb0f2284c/docs/_config.yml#L10-L20

which it required

Alkarex replied:
It would be good to skip the build altogether when the content of `./docs/` has not changed (e.g. based on a hash), which is the case for most PRs / commits

Inverle replied:
Yes, there is a way to do that too. And I thought about it even before this PR.

Inverle replied:
Done: https://github.com/FreshRSS/FreshRSS/pull/8247/commits/df30aa20d9db02cc40acfd321323b34e6c49e4e9

Inverle replied:
Some changes may be required to the repository settings after this PR is merged, but I'm not sure. (needs testing after merge)

Alkarex commented on docs/README.md:
````diff
@@ -0,0 +1,13 @@
+
+Then serve:
+```console
+bundle exec jekyll serve -H 0.0.0.0 --watch --incremental
````
```suggestion
bundle exec jekyll serve -H 127.0.0.1 --watch --incremental
```

A localhost IP is likely sufficient (and safer) for a local test?

Inverle replied:
I prefer 0.0.0.0, but sure.

Alkarex commented on docs/README.md:
````diff
@@ -0,0 +1,13 @@
+This is the documentation deployed by GitHub Pages.
+
+To build and serve locally, first install the necessary packages:
+```console
````
```suggestion
```sh
```

I suggest using `sh` for such commands. `console` is meant to be used with a console prefix such as `$` to distinguish between commands and outputs.

Example with `console` (and without the expected prefix) in VSCode:
<img width="617" height="173" alt="Image" src="https://github.com/user-attachments/assets/4b60c03a-cc2c-4aca-9ec5-9abdd1e860d7" />

Example with `sh` in VSCode:
<img width="541" height="180" alt="Image" src="https://github.com/user-attachments/assets/0281e4af-76d0-4434-8d09-ded51bff5d71" />

Inverle replied:
Good to know, thanks.

Alkarex commented on docs/README.md:
````diff
@@ -0,0 +1,13 @@
+bundle exec jekyll serve -H 0.0.0.0 --watch --incremental
+```
+
+The documentation should be reachable at `http://0.0.0.0:4000/FreshRSS/`.
````
```suggestion
The documentation should be reachable at <http://127.0.0.1:4000/FreshRSS/>.
```

Suggest standard markdown syntax for links

Inverle committed:
Suggestions

Inverle committed:
Use `ruby/setup-ruby` action in workflow for installing and caching gems.

Inverle committed:
Run build only when there are changes to `docs/`

See: https://docs.github.com/en/actions/reference/workflows-and-actions/events-that-trigger-workflows?versionId=free-pro-team%40latest&productId=actions#running-your-workflow-only-when-a-push-to-specific-branches-occurs

Inverle committed:
Merge branch 'edge' into docs-search

Inverle committed:
Change font to `Open Sans`

Inverle commented:
I changed the font to Open Sans

Alkarex commented:
Minor comment: I think line height should be slightly increased:

<img width="865" height="60" alt="image" src="https://github.com/user-attachments/assets/8f66e666-d9eb-41bf-aa7d-7fd78ffb39df" />

Inverle committed:
Increase line height

Inverle committed:
Fix Liquid syntax error

Inverle commented:
@Alkarex done

math-GH commented:
Kudos for this PR!

Inverle commented:
Looks like there was already a sidebar before 👀 
https://web.archive.org/web/20170714232519/http://doc.freshrss.org/en/
